### DISCUSSION TITLE
Migrating to React Native 0.76.x with the New Architecture

### DISCUSSION THREAD

**Discussion by c100k** [zoontek/react-native-bootsplash#657] [Q&A] [answered]
Hey,

Currently migrating my apps to RN 0.76.x with the New Architecture, I'm encountering build issues on iOS with `react-native-bootsplash`.

I opted for a Discussion instead of an Issue because the problem seems to be on our side and not in the lib, since the example seems to compile fine.

![Screenshot 2024-11-27 at 16 15 15](https://github.com/user-attachments/assets/aea95586-624c-4c62-b906-deaea55d368a)

But given the multiple issues reported with the library, I think we all have something in common that is probably misconfigured, but after spending a day or so investigating, I've found nothing. So maybe together we can figure it out by narrowing the issue.

The issues I've found while doing some research :

- 'optional' file not found : https://github.com/facebook/react-native/issues/41872 by @Ananthakr-Jupiter
- iOS build failure react-native v0.76 : https://github.com/zoontek/react-native-bootsplash/issues/651 by @bhandanyan-nomad
- cassert.h, iosfwd.h and memory.h file not found : https://github.com/facebook/react-native/issues/43319 (similar error but not related to the lib)

On my side :

```sh
ScanDependencies /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/AppDelegate.o /MyApp/ios/MyApp/AppDelegate.mm normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler (in target 'MyApp' from project 'MyApp')
    cd /MyApp/ios
    
    Using response file: /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/e6072d4f65d7061329687fe24e3d63a7-common-args.resp
    
    builtin-ScanDependencies -o /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/AppDelegate.o.scan -- /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x objective-c -ivfsstatcache /Users/me/Library/Developer/Xcode/DerivedData/SDKStatCaches.noindex/iphonesimulator18.1-22B74-3d93aac3a03ebac1dd8474c5def773dc.sdkstatcache -fmessage-length\=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit\=0 -fno-color-diagnostics -fmodules-prune-interval\=86400 -fmodules-prune-after\=345600 -fbuild-session-file\=/Users/me/Library/Developer/Xcode/DerivedData/ModuleCache.noindex/Session.modulevalidation -fmodules-validate-once-per-build-session -Wnon-modular-include-in-framework-module -Werror\=non-modular-include-in-framework-module -Wno-trigraphs -Wno-missing-field-initializers -Wno-missing-prototypes -Werror\=return-type -Wunreachable-code -Wno-implicit-atomic-properties -Werror\=deprecated-objc-isa-usage -Wno-objc-interface-ivars -Werror\=objc-root-class -Wno-arc-repeated-use-of-weak -Wimplicit-retain-self -Wduplicate-method-match -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wconditional-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wno-float-conversion -Wnon-literal-null-conversion -Wobjc-literal-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -Wno-selector -Wno-strict-selector-match -Wundeclared-selector -Wdeprecated-implementations -Wno-implicit-fallthrough -fstrict-aliasing -Wprotocol -Wdeprecated-declarations -Wno-sign-conversion -Winfinite-recursion -Wcomma -Wblock-capture-autoreleasing -Wstrict-prototypes -Wno-semicolon-before-method-body -index-store-path /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Index.noindex/DataStore -Wno-comma -Wno-shorten-64-to-32 @/Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/e6072d4f65d7061329687fe24e3d63a7-common-args.resp -MMD -MT dependencies -MF /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/AppDelegate.d --serialize-diagnostics /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/AppDelegate.dia -c /MyApp/ios/MyApp/AppDelegate.mm -o /Users/me/Library/Developer/Xcode/DerivedData/MyApp-ckjhndyflurfvfgyxfozkhvaawms/Build/Intermediates.noindex/MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/AppDelegate.o -index-unit-output-path /MyApp.build/Debug-iphonesimulator/MyApp.build/Objects-normal/x86_64/AppDelegate.o

/MyApp/ios/Pods/Headers/Public/RCTRequired/RCTRequired/RCTRequired.h:8:10: error: 'utility' file not found (in target 'MyApp' from project 'MyApp')
    /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:9: note: while building module 'ReactCodegen' imported from /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:
    /MyApp/ios/<module-includes>:1:9: note: in file included from <module-includes>:1:
    /MyApp/ios/Pods/Headers/Public/ReactCodegen/ReactCodegen-umbrella.h:13:9: note: in file included from /MyApp/ios/Pods/Headers/Public/ReactCodegen/ReactCodegen-umbrella.h:13:
    /MyApp/ios/Pods/Headers/Public/ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h:23:9: note: in file included from /MyApp/ios/Pods/Headers/Public/ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h:23:
/MyApp/ios/Pods/Headers/Public/RCTTypeSafety/RCTTypeSafety/RCTConvertHelpers.h:8:9: error: 'optional' file not found (in target 'MyApp' from project 'MyApp')
    /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:9: note: while building module 'ReactCodegen' imported from /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:
    /MyApp/ios/Pods/Headers/Public/ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h:24:9: note: while building module 'RCTTypeSafety' imported from /MyApp/ios/Pods/Headers/Public/ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h:24:
    /MyApp/ios/<module-includes>:1:9: note: in file included from <module-includes>:1:
    /MyApp/ios/Pods/Headers/Public/RCTTypeSafety/RCTTypeSafety-umbrella.h:13:9: note: in file included from /MyApp/ios/Pods/Headers/Public/RCTTypeSafety/RCTTypeSafety-umbrella.h:13:
/MyApp/ios/Pods/Headers/Public/React-jsi/jsi/jsi.h:10:10: error: 'cassert' file not found (in target 'MyApp' from project 'MyApp')
    /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:9: note: while building module 'ReactCodegen' imported from /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:
    /MyApp/ios/Pods/Headers/Public/React-NativeModulesApple/ReactCommon/RCTTurboModule.h:16:9: note: while building module 'ReactCommon' imported from /MyApp/ios/Pods/Headers/Public/React-NativeModulesApple/ReactCommon/RCTTurboModule.h:16:
    /MyApp/ios/<module-includes>:1:9: note: in file included from <module-includes>:1:
    /MyApp/ios/Pods/Headers/Public/ReactCommon/ReactCommon-umbrella.h:13:9: note: in file included from /MyApp/ios/Pods/Headers/Public/ReactCommon/ReactCommon-umbrella.h:13:
    /MyApp/ios/Pods/Headers/Public/ReactCommon/react/bridging/Array.h:10:10: note: in file included from /MyApp/ios/Pods/Headers/Public/ReactCommon/react/bridging/Array.h:10:
    /MyApp/ios/Pods/Headers/Public/ReactCommon/react/bridging/Base.h:10:10: note: in file included from /MyApp/ios/Pods/Headers/Public/ReactCommon/react/bridging/Base.h:10:
/MyApp/ios/Pods/Headers/Public/glog/glog/logging.h:42:10: error: 'iosfwd' file not found (in target 'MyApp' from project 'MyApp')
    /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:9: note: while building module 'ReactCodegen' imported from /MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:
    /MyApp/ios/Pods/Headers/Public/React-debug/react/debug/react_native_assert.h:54:10: note: while building module 'glog' imported from /MyApp/ios/Pods/Headers/Public/React-debug/react/debug/react_native_assert.h:54:
    /MyApp/ios/<module-includes>:1:9: note: in file included from <module-includes>:1:
    /MyApp/ios/Pods/Headers/Public/glog/glog-umbrella.h:13:9: note: in file included from /MyApp/ios/Pods/Headers/Public/glog/glog-umbrella.h:13:
/MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:9: error: Could not build module 'ReactCodegen' (in target 'MyApp' from project 'MyApp')
    /MyApp/ios/MyApp/AppDelegate.mm:6:9: note: in file included from /MyApp/ios/MyApp/AppDelegate.mm:6:

/MyApp/ios/Pods/Headers/Public/RCTRequired/RCTRequired/RCTRequired.h:8:10: 'utility' file not found

/MyApp/ios/Pods/Headers/Public/RCTTypeSafety/RCTTypeSafety/RCTConvertHelpers.h:8:9: 'optional' file not found

/MyApp/ios/Pods/Headers/Public/React-jsi/jsi/jsi.h:10:10: 'cassert' file not found

/MyApp/ios/Pods/Headers/Public/glog/glog/logging.h:42:10: 'iosfwd' file not found

/MyApp/ios/Pods/Headers/Public/RNBootSplash/RNBootSplash.h:3:9: Could not build module 'ReactCodegen'
```

Multiple times, I've seen that the problem was because of an `.m` file (instead of `.mm`) but that it's not the case on my side. The file is correctly named `AppDelegate.mm`. There is `main.m` but it's the same in the [example](https://github.com/zoontek/react-native-bootsplash/blob/master/example/ios/RNBootSplashExample/main.m).

Also, cleaning the build folder like recommended in https://github.com/zoontek/react-native-bootsplash/issues/651 didn't help neither in my case.

Let's try to find out :)

**Comment by zoontek** [comment]
@c100k If you can reduce the application to the strict minimum and ship a reproduction, I will probably be able to understand what goes wrong here.

Meanwhile, there's not much we can do.

**Reply by c100k** [reply]
You're right, that's a good exercise. I managed to reproduce the issue with a minimal repo : https://github.com/c100k/rn-bootsplash-rn-0.76.x-build-issues-repro. And the culprit seems to be `react-native-reanimated` for absolutely unknown reasons.

When I add the dependency, the build [fails](https://github.com/c100k/rn-bootsplash-rn-0.76.x-build-issues-repro/actions/runs/12072399321). When I remove it, it [succeeds](https://github.com/c100k/rn-bootsplash-rn-0.76.x-build-issues-repro/actions/runs/12072224475).

At the beginning I suspected a mismatch in `CLANG_CXX_LANGUAGE_STANDARD`. `react-native-reanimated` is using `c++17`, most of the other libraries are using `c++20` and `react-native-bootsplash` is the only one using `gnu++0x`. But setting all of them to `c++20` manually didn't solve the problem.

Then I compared the `project.pbxproj` of the `example` with my `project.pbxproj` and they are absolutely identical so it's not from here neither. This is a super weird issue.

**Reply by zoontek** [reply]
I doubt this is an issue. `CLANG_CXX_LANGUAGE_STANDARD` is not enforced in the project podspec, so the used config is the same as everywere else.

![Screenshot 2024-11-29 at 12 44 19](https://github.com/user-attachments/assets/01d3ff93-70f1-4870-b315-f23c91db2cf3)

I can reproduce with the provided repository, thanks, that will help a lot!
Also tried to see if adding `react-native-reanimated` creates the issue in bootsplash example app - it's not 😅

I think there's something wrong, somewhere in your project that mess with Codegen. Will have a hard time to debug it.

**Reply by c100k** [reply]
I was surprised as well that it worked with the example. Even after upgrading it to RN `0.76.3` and RN CLI `15.1.2`. With that upgrade, both projects are "identical" in terms of deps and config.

I've also suspected that the problem could happen with any other library included in `AppDelegate.mm`. But it turned out it didn't when [including](https://github.com/c100k/rn-bootsplash-rn-0.76.x-build-issues-repro/commit/8dad081947cb459a46c549bcf30557830f63b3af) `react-native-config` for instance.

I'll keep digging during the weekend. Even though it's annoying, it's pretty interesting ^^.

**Reply by zoontek** [reply, marked as answer]
@c100k Found it. It's an issue in your project.

It seems that your project is quite old, because you made the update from RN 0.67 to 0.68, which migrate `AppDelegate.m` to `AppDelegate.mm`.

**BUT**. You were supposed to update your project to change `AppDelegate` type to Objective C++:

![Screenshot 2024-11-29 at 17 09 09](https://github.com/user-attachments/assets/dbc780ea-bc03-493d-9137-980690da6062)

And [you didn't](https://github.com/c100k/rn-bootsplash-rn-0.76.x-build-issues-repro/blob/master/ios/MyApp.xcodeproj/project.pbxproj#L21). Switching from `lastKnownFileType = sourcecode.c.objc` to `lastKnownFileType = sourcecode.cpp.objcpp` fix the issue.

To do this directly in Xcode, switch `AppDelegate.mm` to any type excepted `Default - Objective C++ Source`, then set `Default - Objective C++ Source` back. You will see that it updates your project file.

https://github.com/user-attachments/assets/f0d7307d-1a2e-4059-92fd-8c635923fd2f

It was not a `react-native-bootsplash` issue, any third party libraries imported in `AppDelegate.mm` that require Objective C++ would have fail the same way (`react-native-config` could probably be compiled as ObjC alone).

If you enjoyed the free help / support, consider sponsoring, even a one-time payment. It takes me personal free time and is not linked to the library 🙏

**Reply by c100k** [reply]
This is the issue, you're the best !

The project was indeed initiated with RN 0.67, and each upgrade was thoroughly applied via the Upgrade Helper. It looks like I missed some bits.

Thanks a lot for the great job you're doing. I've just sponsored you, fully deserved. Hopefully I can sponsor even more when [RebootX](https://c100k.eu/p/rebootx) reaches its objectives 🙏🏽.

**Reply by zoontek** [reply]
Thanks a lot for the sponsorship, this means a lot 🙏

**Reply by vladimirivanoviliev** [reply]
For me the issue was fixed after noticing that we are targeting too old version of C++ - 11. After updating to C++ 17 the build passed. I also notice that the react native upgrade helper is now targeting C++20 for the latest version, so this is basically missed update between react native version upgrades, simmilar to the issue that you found in the above project.

```
CLANG_CXX_LANGUAGE_STANDARD = "c++17";
```

**Reply by ramnavan** [reply]
I've the same error and I do have correct appdelegate.mm type. trying to migrate from 0.75 to 0.76 with latest react-native-bootsplash. Alls works fine when new architecture is explicitly turned off. when its on, I get these build errors like 

`Import of module 'glog.glog.log_severity' appears within namespace 'google'`

Tried pretty much everything..

**Reply by vladimirivanoviliev** [reply]
@ramnavan The bootsplash package is not the issue for sure. We are on latest 0.79.2 with new arch enabled and it works as expected.

**Comment by code-by** [comment]
I cloned example from this repo and add `RCT_NEW_ARCH_ENABLED=1` into `Preprocessor macros`. 
<img width="791" height="308" alt="image" src="https://github.com/user-attachments/assets/cad88023-cd89-49fd-9c97-bf1556ca4330" />


After that I got errors
```
/react-native-bootsplash/example/ios/RNBootSplashExample/AppDelegate.swift:5:8: Could not build Objective-C module 'RNBootSplash'

/react-native-bootsplash/example/ios/Pods/Headers/Public/ReactCodegen/RNBootSplashSpec/RNBootSplashSpec.h:15:2: This file must be compiled as Obj-C++. If you are importing it, you must change your file extension to .mm.
```

`


<img width="865" height="256" alt="image" src="https://github.com/user-attachments/assets/251becef-a3af-4e69-817a-3b650368a28a" />

so new arch seems not enabled in latest release

<img width="690" height="390" alt="image" src="https://github.com/user-attachments/assets/9a125d95-421d-4582-a295-044cfd582f61" />
